### ISSUE TITLE
Caching discovery information

### ISSUE THREAD

**Issue by andyjeffries** [abonas/kubeclient#487]
Is it possible with the current version of the gem to cache the discovery information for a given server and API set.

We are using Kubernetes as the backing of our API and hence we make a lot of API requests to our API service and then on every one it makes at least ONE call to get service discovery before doing the action.

So I wonder if there's some attribute somewhere I can read and write from/to `Rails.cache` and then no longer need to make that call (during the cache period)?

If not, I'll have a whack at implementing this. I can either then submit it back to the project or maintain it as a monkey-patch locally, would others find it useful in the mainline?

**Comment by andyjeffries** [comment]
BUMP

**Comment by andyjeffries** [comment]
/cc: @cben @astencel-sumo 

Sorry to directly copy you guys, but I don't know if you've subscribed at the repo level so been notified of the new issue. I'm happy to work on this if others would like it but don't have the time to spend on it. I have a major deadline of the 18th but could start after then.

**Comment by cben** [comment]
> We are using Kubernetes as the backing of our API and hence we make a lot of API requests to our API service and then on every one it makes at least ONE call to get service discovery before doing the action.

Wait, every `Client` object should discover only once.  Are you creating a new `Client` for each request?  Can you reuse them?

----

Do you need a cache shared between multiple `Client` objects? Across processes?

If yes, consider also whether you need discovery at all.  What it serves in kubeclient is defining "friendly" method names e.g. `get_pods`, `delete_pod` etc. — derived from discovery listing a `kind: Pod`.
But maybe you don't need that friendliness, and rather than worry about complex caching would rather just construct the correct HTTP request directly?

- In kubeclient 5.0 (and on master since #466, not released yet), we'll have `client.faraday_client` method that will let you make http requests directly.  Not documented yet but intended to become public.  (We also have undocumented `client.rest_client` now in 4.x, but this _is_ going away.)
  If direct HTTP suits your needs, you don't even need `Client`.

- See #332 for plans for a medium-level API.
  My hope there was it might work without discovery at all, but that's tricky due to singular vs plural names :-(, see specifically notes https://github.com/abonas/kubeclient/issues/332#issuecomment-570043664

**Comment by cben** [comment]
[BTW, if your API makes tons of k8s _read_ API calls, perhaps you also be interested in data caching, specifically the "informer" pattern.  Kubeclient doesn't implement it yet (#456) but some other language libraries do.  The idea is List+Watch to initialize a replica of interesting resources in RAM, then keep it updated; then when you need data you just take current value from RAM without k8s request.  This will have some "replication lag" of course, so be careful about consistency with writes.]

**Comment by andyjeffries** [comment]
I may end up going just for a pure HTTP-based client for optimisation. Thanks for the advice.
